File: 003_LongestSubstringWithoutRepeatingCharacters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        currLen, maxLen = 0, 0
        startIndex = 0
        map = dict()
        for i, c in enumerate(s):
            if c not in map:
                map[c] = i
                currLen += 1
            else:
                if currLen > maxLen:
                    maxLen = currLen
                endIndex = map[c]
                temp = endIndex
                for j in range(startIndex, endIndex + 1):
                    del map[s[j]]
                startIndex = temp + 1
                map[c] = i
                currLen = i - temp
        return max(currLen, maxLen)
		
class Solution:
    # @return an integer
    def lengthOfLongestSubstring(self, s):
        start = maxLength = 0
        usedChar = {}
        
        for i in range(len(s)):
            if s[i] in usedChar and start <= usedChar[s[i]]:
                start = usedChar[s[i]] + 1
            else:
                maxLength = max(maxLength, i - start + 1)

            usedChar[s[i]] = i

        return maxLength

File: 003_LongestSubstringWithoutRepeatingCharacters.cpp
//40.7%
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int maxLen(0), currLen(0);
        int startIndex(0);
		unordered_map<char, int> mapS;
		for(int i = 0; i < s.size(); i++){
			unordered_map<char, int>::iterator ite = mapS.find(s[i]);
			if(ite == mapS.end()){
				mapS[s[i]] = i;
				currLen++;
			}else{
				if(currLen > maxLen){
					maxLen = currLen;
				}
				currLen = i - ite -> second;
				int temp(ite -> second);
				for(int j = startIndex; j <= ite -> second; j++)
					mapS.erase(s[j]);
				mapS[s[i]] = i;
				startIndex = temp + 1;
			}
		}
		return max(currLen, maxLen);
    }
};

int lengthOfLongestSubstring(string s) {
        vector<int> dict(256, -1);
        int maxLen = 0, start = -1;
        for (int i = 0; i != s.length(); i++) {
            if (dict[s[i]] > start)
                start = dict[s[i]];
            dict[s[i]] = i;
            maxLen = max(maxLen, i - start);
        }
        return maxLen;
    }